CMakeLists.txt

```diff
@@ -800,6 +800,24 @@ file(APPEND ${CMAKE_BINARY_DIR}/GIT-BUILD-OPTIONS "NO_GETTEXT='${NO_GETTEXT}'\n"
 file(APPEND ${CMAKE_BINARY_DIR}/GIT-BUILD-OPTIONS "RUNTIME_PREFIX='${RUNTIME_PREFIX}'\n")
 file(APPEND ${CMAKE_BINARY_DIR}/GIT-BUILD-OPTIONS "NO_PYTHON='${NO_PYTHON}'\n")
 
+#Make the tests work when building out of the source tree
+if(NOT ${CMAKE_BINARY_DIR} STREQUAL ${CMAKE_SOURCE_DIR})
+	file(RELATIVE_PATH BUILD_DIR_RELATIVE ${CMAKE_SOURCE_DIR} ${CMAKE_BINARY_DIR}/CMakeCache.txt)
+	string(REPLACE "/CMakeCache.txt" "" BUILD_DIR_RELATIVE ${BUILD_DIR_RELATIVE})
+	#Setting the build directory in test-lib.sh before running tests
+	file(WRITE ${CMAKE_BINARY_DIR}/CTestCustom.cmake
+		"file(STRINGS ${CMAKE_SOURCE_DIR}/t/test-lib.sh GIT_BUILD_DIR_REPL REGEX \"GIT_BUILD_DIR=(.*)\")\n"
+		"file(STRINGS ${CMAKE_SOURCE_DIR}/t/test-lib.sh content NEWLINE_CONSUME)\n"
+		"string(REPLACE \"\${GIT_BUILD_DIR_REPL}\" \"GIT_BUILD_DIR=\\\"$TEST_DIRECTORY\\\"/../${BUILD_DIR_RELATIVE}\" content \"\${content}\")\n"
+		"file(WRITE ${CMAKE_SOURCE_DIR}/t/test-lib.sh \${content})")
```

Comment:
This would leave the file `t/test-lib.sh` modified by the build process, right? I'd rather not do that.

Instead, I would change `t/test-lib.sh` in a separate, preparatory commit (read: _before_ the first commit that creates `CMakeLists.txt`) such that it only sets `GIT_BUILD_DIR` unless it is already set. I.e. [this line](https://github.com/git/git/blob/de49261b050d9cd8ec73842356077bc5b606640f/t/test-lib.sh#L37) would be expanded to:

```sh
# only set GIT_BUILD_DIR unless it is already set.
test -n "$GIT_BUILD_DIR" || GIT_BUILD_DIR="$TEST_DIRECTORY"/..
```

Hopefully we can make sure that that variable is set when running the test scripts via CTest?

Reply:
Not during the build. Only if you choose to run the tests from out of source (i.e call ctest). 
Also found a way of reverting test-lib.sh after the tests complete, but this means the tests should complete (will take 30 mins or so).
We can also set an environment variable, but it should be unix like.
I did try with C:/proj/git/build (intended forward slash) it did not work properly in this case.(Giving absolute paths with forward slashed usually works)

Reply:
Hmm, pity.